test/config/utility.cc

```diff
@@ -93,4 +117,84 @@ void ConfigHelper::setSourceAddress(const std::string& address_string) {
       ->set_address(address_string);
 }
 
+void ConfigHelper::setDefaultHostAndRoute(const std::string& domains, const std::string& prefix) {
+  envoy::api::v2::filter::HttpConnectionManager hcm_config;
+  loadHttpConnectionManager(hcm_config);
+
+  auto virtual_host = hcm_config.mutable_route_config()->mutable_virtual_hosts(0);
+  virtual_host->set_domains(0, domains);
+  virtual_host->mutable_routes(0)->mutable_match()->set_prefix(prefix);
+
+  storeHttpConnectionManager(hcm_config);
+}
+
+void ConfigHelper::setBufferLimits(uint32_t upstream_buffer_limit,
+                                   uint32_t downstream_buffer_limit) {
+  RELEASE_ASSERT(bootstrap_.mutable_static_resources()->listeners_size() == 1);
+  auto* listener = bootstrap_.mutable_static_resources()->mutable_listeners(0);
+  listener->mutable_per_connection_buffer_limit_bytes()->set_value(downstream_buffer_limit);
+
+  auto* static_resources = bootstrap_.mutable_static_resources();
+  for (int i = 0; i < bootstrap_.mutable_static_resources()->clusters_size(); ++i) {
+    auto* cluster = static_resources->mutable_clusters(i);
+    cluster->mutable_per_connection_buffer_limit_bytes()->set_value(upstream_buffer_limit);
+  }
+}
+
+void ConfigHelper::addRoute(const std::string& domains, const std::string& prefix,
+                            const std::string& cluster,
+                            envoy::api::v2::VirtualHost::TlsRequirementType type) {
+  envoy::api::v2::filter::HttpConnectionManager hcm_config;
+  loadHttpConnectionManager(hcm_config);
+
+  auto virtual_host = hcm_config.mutable_route_config()->add_virtual_hosts();
+  virtual_host->add_domains(domains);
+  virtual_host->add_routes()->mutable_match()->set_prefix(prefix);
+  virtual_host->mutable_routes(0)->mutable_route()->set_cluster(cluster);
+  virtual_host->set_require_tls(type);
+
+  storeHttpConnectionManager(hcm_config);
+}
+
+void ConfigHelper::addFilter(const std::string& config) {
+  envoy::api::v2::filter::HttpConnectionManager hcm_config;
+  loadHttpConnectionManager(hcm_config);
+
+  auto* filter_list_back = hcm_config.add_http_filters();
+  const std::string json = Json::Factory::loadFromYamlString(config)->asJsonString();
+  MessageUtil::loadFromJson(json, *filter_list_back);
+
+  // Now move it to the front.
+  for (int i = hcm_config.http_filters_size() - 1; i > 0; --i) {
+    hcm_config.mutable_http_filters()->SwapElements(i, i - 1);
```

Comment:
Why not make the API work in the other direction?

Reply:
Do you mean have envoy load front to back, or have addFilter add to the back?

Generally we start with a router filter so adding things behind it is not so useful.